packages/asset-swapper/src/utils/market_operation_utils/index.ts

```diff
@@ -240,373 +198,114 @@ export class MarketOperationUtils {
-                optimalPath = path;
-                optimalPathOutput = output;
-            }
+    private _generateOptimizedOrders(opts: {
+        side: MarketOperation;
+        inputToken: string;
+        outputToken: string;
+        inputAmount: BigNumber;
+        nativeOrders: SignedOrder[];
+        orderFillableAmounts: BigNumber[];
+        dexQuotes: DexSample[][];
+        ethToOutputRate?: BigNumber;
+        bridgeSlippage?: number;
+        excludedSources?: ERC20BridgeSource[];
+        fees?: { [source: string]: BigNumber };
+        allowFallback?: boolean;
+        liquidityProviderAddress?: string;
+    }): OptimizedMarketOrder[] {
+        const { inputToken, outputToken, side, inputAmount } = opts;
+        // Convert native orders and dex quotes into fill paths.
+        const paths = createFillPaths({
+            side,
+            // Augment native orders with their fillable amounts.
+            orders: createSignedOrdersWithFillableAmounts(side, opts.nativeOrders, opts.orderFillableAmounts),
+            dexQuotes: opts.dexQuotes,
+            ethToOutputRate: opts.ethToOutputRate,
+            excludedSources: opts.excludedSources,
+            fees: opts.fees,
+        });
+        // Find the optimal path.
+        const optimalPath = findOptimalPath(side, paths, inputAmount);
+        // console.log(JSON.stringify(paths, null, '  '));
+        if (!optimalPath) {
+            throw new Error(AggregationError.NoOptimalPath);
         }
-    }
-    return optimalPath;
-}
-
-// Gets the total input of a path.
-function getPathInput(path: Fill[]): BigNumber {
-    return BigNumber.sum(...path.map(p => p.input));
-}
-
-// Merges contiguous fills from the same DEX.
-function collapsePath(path: Fill[], isBuy: boolean): CollapsedFill[] {
-    const collapsed: Array<CollapsedFill | NativeCollapsedFill> = [];
-    for (const fill of path) {
-        const makerAssetAmount = isBuy ? fill.input : fill.output;
-        const takerAssetAmount = isBuy ? fill.output : fill.input;
-        const source = (fill.fillData as FillData).source;
-        if (collapsed.length !== 0 && source !== ERC20BridgeSource.Native) {
-            const prevFill = collapsed[collapsed.length - 1];
-            // If the last fill is from the same source, merge them.
-            if (prevFill.source === source) {
-                prevFill.totalMakerAssetAmount = prevFill.totalMakerAssetAmount.plus(makerAssetAmount);
-                prevFill.totalTakerAssetAmount = prevFill.totalTakerAssetAmount.plus(takerAssetAmount);
-                prevFill.subFills.push({ makerAssetAmount, takerAssetAmount });
-                continue;
-            }
+        // Find a fallback path from sources not used in the first path.
+        let fallbackPath: Fill[] = [];
+        if (opts.allowFallback) {
+            fallbackPath = findOptimalPath(side, getUnusedSourcePaths(optimalPath, paths), inputAmount) || [];
```

Comment:
I believe this would pickup Kyber where Uniswap was previously in the Optimal Path. These would be exclusive as if the Uniswap path were to fail then the Kyber path would probably fail, at least at that price? 

Seems risky.